### ISSUE TITLE
Anything standing against merging with Distrobox?

### ISSUE THREAD

**Issue by trytomakeyouprivate** [containers/toolbox#1272]
**Is your feature request related to a problem? Please describe.**
Distrobox is based off Toolbox, but offers a lot more functionalities, that are partly even open issues in toolbox.

- both use Podman
- both work with the same commands
- both by default create a Fedora container
- Distrobox adds .desktop items for apps and Containers
- Distrobox has way more tested Distro images

**Describe the solution you'd like**
Just merge with Distrobox, why this duplicate efford? I am pretty sure it would benefit both systems, as they are pretty much the same. I dont know any advantages of Toolbox over Distrobox, but there are probably some. 

Both have done work that would help the other project.

**Describe alternatives you've considered**
At the moment I am just annoyed with Fedora Kinoite shipping Toolbox and I replace it with Distrobox, which is sad.

**Comment by akdev1l** [comment]
Toolbox and distrobox are very different such "merge" would probably involve letting one of the two projects die

Distrobox has issues too generally it is more janky ime and the images aren't really "tested" as much as they hack around to make them work but an update could break it. 

checkout https://ublue.it - for a Fedora Kinoite based image that comes with distrobox by default

**Comment by trytomakeyouprivate** [comment]
Hi, yes I see that point. Currently Toolbox supports the images I need, so this could be working. Still I like the amount of images in Distrobox.

Is the software itself that different? Does Toolbox support exporting apps and binaries to the system for native integration? This is a really nice feature. But I see why Fedora would choose something more stable.

**Comment by iamcult** [comment]
iirc distrobox is written completely in bash, while toolbox is written in go. afaik distrobox is in no way "based" on toolbox, and has an independent codebase.

**Comment by iamcult** [comment]
it might help to mention that you can specify a custom image remote in toolbox currently, it just doesn't include a list of remotes internally like distrobox.

**Comment by trytomakeyouprivate** [comment]
Interesting! There are no Discussions in this repo, so I would keep this issue open for discussion.

It seems at least for managing images Toolbox can work the same then.

Any other important changes, making one "better", that you know of?

**Comment by akdev1l** [comment]
my full opinion on this is that toolbx is taken a better/cleaner approach(well tested images with well defined requirements) but sadly this isn't really reflected in the codebase and the state of the project seems dubious (disclaimer: I've got a PR that is over a year old that tries to fix a small issue that is a few years old so I am very biased)

on the other hand distrobox takes an approach that I think isn't workable and gives less consistent results (trying to hack around the different container images to support "everything") but the codebase is far easier to understand and the project seems to be very active when it comes to new features and fixes. 

I am not really happy with either but at the moment I use distrobox. 

(just a note here to clarify that I mean no disrespect to any of the projects - I've learned a lot by observing them and studying them)

**Comment by debarshiray** [comment]
It was a deliberate decision to *rewrite* Toolbx in Go, and @HarryMichal did all the heavy lifting here.  It used to be in POSIX shell before that, and I want to avoid going back to a shell script implementation again.

Note that this isn't a trivial change.  The `toolbox(1)` executable from the host is used inside the Toolbx container, which has all sorts of ABI compatibility concerns when the executable is a compiled binary instead of an interpreted script.

**Comment by debarshiray** [comment]
Today there's built-in support for [Arch Linux](https://github.com/containers/toolbox/commit/2ee82affebd49a86bdc43517d2ffb6dc19c6784c) and [Ubuntu](https://github.com/containers/toolbox/commit/a84a358b3b90dac9a66f6034434843a612aa888f), just like we have had for Fedora and Red Hat Enterprise Linux for a while now.

**Comment by debarshiray** [comment]
> Distrobox is based off Toolbox, but offers a lot more functionalities,
> that are partly even open issues in toolbox.

Yes, it's true that Distrobox offers a lot more functionalities than Toolbx.  In some cases, it's a deliberate non-goal for Toolbx to offer some of those.  For example:

>     * both use Podman

Distrobox also supports Docker.  You can use the `DBX_CONTAINER_MANAGER` environment variable to switch between Docker and Podman.

We are unlikely to do that in Toolbx unless it unlocks a substantial user-visible feature that would otherwise be impossible with Podman.  Working on Toolbx also means owning the Podman stack, which is why Toolbx is part of the same Containers organization on GitHub as Podman.  This includes driving new features, and tracking down regressions and bugs, which is a very significant effort.  Doing that for another container management stack, without any substantial user-visible benefits, isn't worth it.

Note that *also works with Docker* isn't a substantial user-visile feature.  :)

>     * Distrobox adds .desktop items for apps and Containers

This is another non-goal for Toolbx.  We consider [Flatpak](https://flatpak.org/) to be the solution for graphical client-side applications.  Toolbx is meant to complement Flatpak, not replace it.  Toolbx does offer the `run` command for temporary workarounds, but that doesn't mean that it should get into the business of handling `*.desktop` files.  Flatpak already does that.

Instead we want to solve prooblems like enablng IDEs such as Visual Studio Code, shipped as Flatpaks, to work with Toolbx containers.

**Comment by debarshiray** [comment]
Closing.  There is nothing actionable here, and the discussion seems to have died down.

**Comment by trytomakeyouprivate** [comment]
Agree with many points and learned a lot.

But it is just not true that Flatpak works for everything. Its useless and way too complex for little command line tools, where `distrobox-export --bin /path/to/bin` is really useful. The same goes for some rare apps like RStudio that use the systems package manager, which doesnt work on Fedora Atomic and others.

It is a nieche feature, but really worth doing. Without it, I would never use Toolbox, as Distrobox just works.

**Comment by debarshiray** [comment]
> But it is just not true that Flatpak works for everything. Its useless and way too complex for little command line tools

What sort of command line tools are we talking about?  Is it something that's already available through traditional packages like RPMs?  Or is it something custom or a one-off throwaway experiment?

There does [exist](https://github.com/flathub/org.flatpak.Builder) a Flatpak for `flatpak-builder`.  So, it's not entirely useless for command line tools, even though that's not what it's designed for.

> where `distrobox-export --bin /path/to/bin` is really useful.

You can replicate that with `alias mycmd='toolbox run mycmd'` when `/run/.containerenv` is absent using the shell's start-up files.

That doesn't seem that much more complicated if the CLI tool is already available through a RPM package or similar.  Else, if you have to locally build or otherwise set up the tool, then that's the more complicated part.

Or am I missing something?

> The same goes for some rare apps like RStudio that use the systems package manager, which doesnt work on Fedora Atomic and others.

See my comment about Visual Studio Code above.  Also:
* https://discussion.fedoraproject.org/t/toolbox-and-visual-studio-code-remote-containers/27987
* https://github.com/owtaylor/toolbox-vscode

**Comment by trytomakeyouprivate** [comment]
@debarshiray examples: cheat, tldr, bat, cargo, pip, dnf (searching for Fedora packages)... Yes these are RPMs, what else? If it werent RPMs you could just store the binaries in your $PATH and probably avoid containers. But RPMs (or other Distro packages) are easier to setup including dependencies etc.

Yes of course you can alias the commands manually, all Distrobox does is to do this automatically for you, but not with an alias (only working for a specific shell, or needing shell detection, or duplicating) but with a script in your $PATH

You dont have to locally build anything, did you use Distrobox? Its pretty much the same:
- podman container automatically created
- use different container images for Distros
- Install native packages into that container using the Distros pac

Distrobox uses different code, but does
- desktop entries for the Containers 
- easy setup of desktop entries of GUI apps, example `~/.local/share/applications/Fedora-rstudio.desktop`:

```
[Desktop Entry]
Exec=/usr/bin/distrobox-enter  -n Fedora --   env QT_QPA_PLATFORM=xcb rstudio  %F
Icon=rstudio
Type=Application
Terminal=false
Name=RStudio  (on Fedora)
Categories=Development;IDE;
MimeType=text/x-r-source;text/x-r;text/x-R;text/x-r-doc;text/x-r-sweave;text/x-quarto-markdown;text/x-r-markdown;text/x-r-html;text/x-r-presentation;application/x-r-data;application/x-r-project;application/x-rdp-rsp;text/x-r-history;text/x-r-profile;text/x-tex;text/x-markdown;text/css;text/javascript;text/x-chdr;text/x-csrc;text/x-c++hdr;text/x-c++src;
StartupWMClass=rstudio
```

- easy setup of scripts in your PATH, example `~/.bin/cheat`:

```
#!/bin/sh
# distrobox_binary
# name: Fedora
if [ ! -f /run/.containerenv ] && [ ! -f /.dockerenv ]; then
	command="/usr/bin/distrobox-enter  -n Fedora --  /usr/bin/cheat  "

	for arg in "$@"; do
		if echo "${arg}" | grep -Eq "'|\""; then
			command="${command} \
				$(echo "${arg}" | sed 's|\\|\\\\|g' |
				sed 's| |\\ |g' |
				sed 's|\$|\\\$|g' |
				sed "s|'|\\\'|g" |
				sed 's|"|\\\"|g')"
		elif echo "${arg}" | grep -q "'"; then
			command="${command} \"${arg}\""
		else
			command="${command} '${arg}'"
		fi
	done
	eval ${command}
else
	/usr/bin/cheat "$@"
fi
```

These are pretty impressive and very useful. 

I see Toolbox supporting more and more container images, but without these features there is no real reason to use it over Distrobox.

**Comment by debarshiray** [comment]
> Yes of course you can alias the commands manually, all Distrobox
> does is to do this automatically for you, but not with an alias (only
> working for a specific shell, or needing shell detection, or duplicating)
> but with a script in your $PATH

How often do you change shells and what shells do you use?  Aliases interoperate across at least Bash and Z shell.

I don't see how setting up a shell alias versus having a separate Toolbx command to do it for you makes a big difference.  Unless these aliases are constantly being changed, which would be surprising, because it sounds like something one would set up once and keep using over time.

> You dont have to locally build anything,

I meant locally building code that's not shipped as an RPM.

>     * easy setup of scripts in your PATH, example `~/.bin/cheat`:
> 
> 
> ```
> #!/bin/sh
> # distrobox_binary
> # name: Fedora
> if [ ! -f /run/.containerenv ] && [ ! -f /.dockerenv ]; then
> 	command="/usr/bin/distrobox-enter  -n Fedora --  /usr/bin/cheat  "
> 
> 	for arg in "$@"; do
> 		if echo "${arg}" | grep -Eq "'|\""; then
> 			command="${command} \
> 				$(echo "${arg}" | sed 's|\\|\\\\|g' |
> 				sed 's| |\\ |g' |
> 				sed 's|\$|\\\$|g' |
> 				sed "s|'|\\\'|g" |
> 				sed 's|"|\\\"|g')"
> 		elif echo "${arg}" | grep -q "'"; then
> 			command="${command} \"${arg}\""
> 		else
> 			command="${command} '${arg}'"
> 		fi
> 	done
> 	eval ${command}
> else
> 	/usr/bin/cheat "$@"
> fi
> ```

Just a note: this particular wrapper script with it's huge number of `sed(1)` invocations in a loop and an invalid negation in the `if` branch isn't the best way to strengthen your argument.

Maybe use the very different looking script that recent versions of Distrobox generate.  :)

> These are pretty impressive and very useful.
> 
> I see Toolbox supporting more and more container images, but without these features there is no real reason to use it over Distrobox.

That's alright.  I am not trying to convince you to use Toolbx.

**Comment by trytomakeyouprivate** [comment]
@debarshiray I use fish for everything, bash for running scripts.

Especially the Distrobox GUI desktop entry exports are simply way better and work automatically without the script path.

No way of that exact wrapper script but if its tested it may deal with issues I cant really recognise and in 99% of cases it will simply use the second option in the if loop.

No YOU dont have to convince ME of distrobox. I will just continue using Toolbox.

But along with Fedora Flatpaks, toolbox is the one thing that is just annoying and that I always have to change. Distrobox has the way better UX for non-coders. Its the best CLI of both.

- automatic aliases in the PATH so they work in bash, zsh and fish
- easy GUI desktop entries
- lots of supported images

these are simply better. The rest is arguable.

**Comment by trytomakeyouprivate** [comment]
@debarshiray I use fish for everything, bash for running scripts.

Especially the Distrobox GUI desktop entry exports are simply way better and work automatically without the script path.

No way of that exact wrapper script but if its tested it may deal with issues I cant really recognise and in 99% of cases it will simply use the second option in the if loop.

No YOU dont have to convince ME of distrobox. I will just continue using Toolbox.

But along with Fedora Flatpaks, toolbox is the one thing that is just annoying and that I always have to change. Distrobox has the way better UX for non-coders. Its the best CLI of both.

- automatic aliases in the PATH so they work in bash, zsh and fish
- easy GUI desktop entries
- lots of supported images

these are simply better. The rest is arguable. I am not arguing personally, but Fedora people refuse to ship Distrobox, Toolbox sucks for UX so I have to say "Fedora is great, if you change XYZ" and thats not what a good distro should be like

**Comment by debarshiray** [comment]
*I took the liberty to delete the duplicated comment.*

> @debarshiray I use fish for everything, bash for running scripts.

fish supports the same syntax for aliases as Bash and Z shell.

> Especially the Distrobox GUI desktop entry exports are simply way better and work automatically without the script path.
> 
> No way of that exact wrapper script but if its tested it may deal with issues I cant really recognise

I didn't understand this part.

> and in 99% of cases it will simply use the second option in the if loop.

There's nothing *simple* about that maze of `sed(1)` invocations.  There's a reason Distrobox no longer uses that.  :)

Anyway, if you are able to read and understand that, then you can surely put this in your start-up scripts:
```
if ! [ -f /run/.containerenv ]; then
    alias mycmd='toolbox run mycmd'
fi
```

Or am I missing something?

Aliases are a common and relatively easy-to-use feature of shells.  I am hesitant to replicate that separately in Toolbx unless there's a clear advantage.

Every extra feature in Toolbx means extra lines of code that need to be maintained in backwards compatible ways and extra tests.  We have close to 300 tests that get run across CentOS Stream 9, Fedora and Ubuntu, and it takes a non-trivial amount of time to run.  That directly translates to things like power consumption and contributor time.

So, once again, I don't want to add features to accomplish things that can already be achieved with comparable effort.

> No YOU dont have to convince ME of distrobox. I will just continue using Toolbox.

That would be lovely, but I don't think you meant that.  :)
 
> But along with Fedora Flatpaks, toolbox is the one thing that is just annoying and that I always have to change. Distrobox has the way better UX for non-coders. Its the best CLI of both.
> 
>     * automatic aliases in the PATH so they work in bash, zsh and fish

There's nothing *automatic* about it in Distrobox either, because you have to invoke `distrobox export`.

It seems that you prefer to invoke `distrobox export` and I am saying that the same can be achieved by shell aliases.  Maybe there is something I am missing?

>     * easy GUI desktop entries
> 
>     * lots of supported images

This *lots of supported images* claim is invalid.

It has always been possible to specify any custom image with `toolbox create --image`, which is exactly what Distrobox also does.

Some of the images that are listed on [distrobox/docs/compatibility](https://github.com/89luca89/distrobox/blob/main/docs/compatibility.md) are actually maintained as part of the coreeToolbx project (ie., the images for Arch Linux, Fedora, RHEL and Ubuntu).  Many come from the satellite [github.com/toolbx-images/images](https://github.com/toolbx-images/images) project.

Toolbx advertises a small subset of operating system distributions as supported and makes it easier to create containers for them through the `--distro` option.  Currently, this is Arch Linux, Fedora, RHEL and Ubuntu.  This is because:
* We need a dedicated maintainer from that specific distribution who is willing to take care of distro-specific problems, and drive changes in other parts of the distribution if needed.
* Ideally be able to run our CI on hosts (not containers) with the distribution, so that we are sure that things work.

If someone wants to have their distribution supported upstream then they need to sign up for this work.

If they don't want to do this work upstream, for whatever reason, then they can ship a `toolbox.conf(5)` downstream in their distribution that uses the `image` option to specify a default image for their distribution.   This way `toolbox create` (without any options) won't fallback to some other image and will use this specified image.

So, yes, this claim of *lots of supported images* is invalid.

> these are simply better. The rest is arguable. I am not arguing personally but Fedora people refuse to ship Distrobox,

Fedora [has](https://src.fedoraproject.org/rpms/distrobox) a Distrobox RPM.

You do realize that Fedora contributors do a LOT of heavy lifting for Toolbx?  Distrobox gets this work for free.

> Toolbox sucks for UX so I have to say "Fedora is great, if you change XYZ" and thats not what a good distro should be like

That's your opinion and that's fine.

As a long-time Fedora contributor, I disagree with you.  So, just throwing your opinion with phrases like *this sucks* and *this is annoying* isn't going to go anywhere.

**Comment by trytomakeyouprivate** [comment]
I see that their method of aliasing is kinda weird, but as I said, in 99% of the time that whole sed blubb will not be used as its only for some arguments.

Sure, every extra line is problematic, maybe. I dont know exactly how they do the GUI entries (okay [somewhere here](https://github.com/89luca89/distrobox/blob/main/distrobox-export)). Their shell aliasing is controversial.

I know fish can use `alias "x"="y"` too but the tool has to determine the used shell(s) and use the corrent config path. You need to know the binary path too, so `distrobox-export --bin` is not the thing I would like to be ported, but `distrobox-export --app`

The GUI app integration really is great. Also, the Distro containers themselves get a graphical Desktop entry automatically, including a nice Distro+Box Icon, which is really nice. So after once installing it, you are GUI, partly. In the GUI you click on it, install your app and export it, and the exported app has a description that it runs in the container, has the correct icon, mimetypes etc. Dont know how they do it, probably simply copy over the desktop entries.

I have no idea of Go, but those two things, automatic container entries, and easy graphical app entries should not really be hard, a handful lines of code I guess. Not demanding this of anyone, but acknowledge that it has a use?

 Okay, got what you mean by "supported images". They are simply there but not supported. I think the new subset of images in Toolb(o?)x is to be preferred, as long as it contains nice stuff like Fedora 39, rawhide etc. Because I dont need tested images if I want to test myself. This is a bit different, a way in between, like some way to enable custom images would be really great.

Okay, "ships" means "preinstalls" for me. Especially on Atomic desktops, its a clear decision.

Yes I see that Toolbox contributors create and maintain images. But there are a lot more too, yes they are not supported okay, but see, these are not all created by Fedora Developers:

```
[user@fedora user]$ distrobox-create -i 
cgr.dev/chainguard/wolfi-base:latest                      docker.io/linuxdeepin/beige                               quay.io/toolbx-images/centos-toolbox:stream9
container-registry.oracle.com/os/oraclelinux:7            docker.io/linuxmintd/mint21.1-amd64                       quay.io/toolbx-images/debian-toolbox:10
container-registry.oracle.com/os/oraclelinux:7-slim       docker.io/openeuler/openeuler:latest                      quay.io/toolbx-images/debian-toolbox:11
container-registry.oracle.com/os/oraclelinux:8            docker.io/vbatts/slackware:14.2                           quay.io/toolbx-images/debian-toolbox:testing
container-registry.oracle.com/os/oraclelinux:8-slim       ghcr.io/void-linux/void-linux:latest-full-x86_64          quay.io/toolbx-images/debian-toolbox:unstable
container-registry.oracle.com/os/oraclelinux:9            ghcr.io/void-linux/void-linux:latest-full-x86_64-musl     quay.io/toolbx-images/opensuse-toolbox:latest
container-registry.oracle.com/os/oraclelinux:9-slim       public.ecr.aws/amazonlinux/amazonlinux:1                  quay.io/toolbx-images/rockylinux-toolbox:8
docker.io/debian/eol:wheezy                               public.ecr.aws/amazonlinux/amazonlinux:2                  quay.io/toolbx-images/rockylinux-toolbox:9
docker.io/gentoo/stage3:latest                            public.ecr.aws/amazonlinux/amazonlinux:2022.0.20220531.0  quay.io/toolbx-images/ubuntu-toolbox:16.04
docker.io/kalilinux/kali-rolling:latest                   quay.io/almalinux/8-base:8                                quay.io/toolbx-images/ubuntu-toolbox:18.04
docker.io/library/alpine:3.15                             quay.io/almalinux/8-init:8                                quay.io/toolbx-images/ubuntu-toolbox:20.04
docker.io/library/alpine:3.16                             quay.io/almalinux/almalinux:8                             quay.io/toolbx-images/ubuntu-toolbox:22.04
docker.io/library/alpine:latest                           quay.io/almalinux/almalinux:9                             registry.access.redhat.com/ubi7/ubi
docker.io/library/archlinux:latest                        quay.io/almalinux/almalinux:9-minimal                     registry.access.redhat.com/ubi7/ubi-init
docker.io/library/clearlinux:base                         quay.io/centos/centos:7                                   registry.access.redhat.com/ubi8/toolbox
docker.io/library/clearlinux:latest                       quay.io/centos/centos:stream8                             registry.access.redhat.com/ubi8/ubi
docker.io/library/debian:10                               quay.io/centos/centos:stream9                             registry.access.redhat.com/ubi8/ubi-init
docker.io/library/debian:stable                           quay.io/fedora/fedora:36                                  registry.access.redhat.com/ubi8/ubi-minimal
docker.io/library/debian:stable-backports                 quay.io/fedora/fedora:38                                  registry.access.redhat.com/ubi9/toolbox
docker.io/library/debian:testing                          quay.io/fedora/fedora:39                                  registry.access.redhat.com/ubi9/ubi
docker.io/library/debian:testing-backports                quay.io/rockylinux/rockylinux:8                           registry.access.redhat.com/ubi9/ubi-init
docker.io/library/debian:unstable                         quay.io/rockylinux/rockylinux:8-minimal                   registry.access.redhat.com/ubi9/ubi-minimal
docker.io/library/mageia                                  quay.io/rockylinux/rockylinux:9                           registry.fedoraproject.org/fedora:37
docker.io/library/neurodebian:nd100                       quay.io/rockylinux/rockylinux:latest                      registry.fedoraproject.org/fedora-toolbox:37
docker.io/library/sl:7                                    quay.io/toolbx-images/almalinux-toolbox:8                 registry.fedoraproject.org/fedora-toolbox:38
docker.io/library/ubuntu:14.04                            quay.io/toolbx-images/almalinux-toolbox:9                 registry.fedoraproject.org/fedora-toolbox:39
docker.io/library/ubuntu:16.04                            quay.io/toolbx-images/alpine-toolbox:3.16                 registry.getcryst.al/crystal/misc/docker:latest
docker.io/library/ubuntu:18.04                            quay.io/toolbx-images/alpine-toolbox:3.17                 registry.opensuse.org/opensuse/distrobox:latest
docker.io/library/ubuntu:20.04                            quay.io/toolbx-images/alpine-toolbox:edge                 registry.opensuse.org/opensuse/leap:latest
docker.io/library/ubuntu:22.04                            quay.io/toolbx-images/archlinux-toolbox                   registry.opensuse.org/opensuse/toolbox:latest
docker.io/linuxdeepin/apricot                             quay.io/toolbx-images/centos-toolbox:stream8              registry.opensuse.org/opensuse/tumbleweed:latest
```

thanks for your time to answer all my questions. In the end I think we could clear up many things, and (sorry that through this medium) I learned a lot.

**Comment by debarshiray** [comment]
> Yes I see that Toolbox contributors create and maintain images.

Not just the images, but the entire Podman stack down to the Linux kernel, fixing various bits and pieces to work well inside containers, all the infrastructure work to serve the images, etc..

> But there are a lot more too, yes they are not supported okay, but see, these are not all created by Fedora Developers:

I don't understand.  Most of those are just generic OCI images that are listed in a [file](https://github.com/89luca89/distrobox/blob/main/docs/compatibility.md).

What's so special about them?

We don't want to install packages into the container at run-time when creating or starting a container.  We get too many bugs where a container didn't start for some non-obvious reason.  Adding package installation to that will make things even more unpredictable and untestable, because we would have no control over what repositories are enabled and what packages are already installed in the images.

Starting from an image and not changing it when creating and starting the container leads to an outcome that's a lot more predictable and testable.

**Comment by debarshiray** [comment]
> The GUI app integration really is great. Also, the Distro containers themselves get a graphical Desktop entry automatically, including a nice Distro+Box Icon, which is really nice. So after once installing it, you are GUI, partly. In the GUI you click on it, install your app and export it, and the exported app has a description that it runs in the container, has the correct icon, mimetypes etc. Dont know how they do it, probably simply copy over the desktop entries.

A lot of those things have significant overlap with [Flatpak](https://flatpak.org/).  Users are better off using Flatpak containers for graphical applications.
 
> I have no idea of Go, but those two things, automatic container entries, and easy graphical app entries should not really be hard, a handful lines of code I guess. Not demanding this of anyone, but acknowledge that it has a use?

It's not the same thing, but I would encourage using [Prompt](https://gitlab.gnome.org/chergert/prompt/) as the terminal emulation application with Toolbx.  It has a lot better Toolbx integration than GNOME Terminal.  eg., you can graphically chose which container to use.

**Comment by boredsquirrel** [comment]
@debarshiray no Flatpak does not cover this.

Flatpak is only meant for stable apps, does not work for IDEs, RStudio, QGis and other apps needing plugin systems with package managers. They dont work for testing apps, running prerelease software and a lot more.

There are also occasions where there is no flatpak. For sure this is not a good scenario, but in the case of immutable Distros, many people need containers.

To the comment above, I dont know where you read that installing packages at creation would be needed.

I see the projects are different, and already opened some issues about features missing compared to distrobox. Until at least these are implemented, Fedora Atomic does not reach its potential.

**Comment by marbub** [comment]
@debarshiray Flatpak support is still not working for IDEs. For example, you want to support different Qt versions for Qt Creator. But Flatpak does not support extensions for different Qt versions. You only have the KDE runtime, which is very KDE-specific.

I like to use Davinci Resolve. You can find a distro box container for it, but no flatpak. The problem with flatpak is simply support. It is 80% done, but the last 20% are missing. And I don't see a big push to change it.

**Comment by lduktus** [comment]
There seems to be still a lot of discussion, so I'd like to add another perspective.

I agree with some of the points: Flatpak is great for many things, but certain apps aren't plug-and-play. And relying on toolbx or distrobox is a hack –– sometimes necessary, but not ideal.

For example, IDE support is unlikely to come quick. While VSCode has a devcontainer feature, Podman support is less seamless and I am not sure if native support for toolboxes is on their agenda. Many other IDEs lack devcontainer concepts, and plugins for apps like Obsidian may require additional tools like Pandoc or Python packages. Another example: There is a request for the Zotero Flatpak to add Tesseract for a specific plugin. I use Zotero in a distrobox due to a Flatpak bug. I used the occasion and installed Tesseract to test the plugin and may stay on this setup for now. Long-term, these kind of issues need solutions to make Flatpaks more viable. These may be edge cases, but a lot of users have their particular edge case.

Toolbx doesn't integrate all features of distrobox, but most can be done with more effort. Distrobox isn't hard to install, so there's no reason not to use both. However a better way to configure toolboxes, similar to distrobox-assemble, would make them more reproducible and easier to manage. Features like mounting additional paths or named volumes would also be useful.

Long-term, broader integration of containerization and virtualization technologies into immutable OSs would be very nice, e. g. incus could be interesting. While this isn't directly related to toolbx, many workflows would benefit from system containers or proper VMs. Using OCI containers as lightweight VMs has its limitations. As far as I know Ubuntu Core OS uses LXD. Imho having different options would be great: toolboxes –– either via toolbx or distrobox –– for tighter Integration CLI tools and feeling like home, devcontainers for certain projects, and system containers for broader environment or if the devcontainer workflow doesn't fit your needs.

Long story short I wonder if we are not having the wrong discussion, when we ask toolbox to be a distrobox replacement or vice versa. This won't help to move the ecosystem and Desktop experience forward.

**Comment by debarshiray** [comment]
> Flatpak is only meant for stable apps, does not work for IDEs, RStudio, QGis and other apps needing plugin systems with package managers.

I already answered this twice before.

Please understand that some of us are busy doing the heavy lifting to make things properly.  Constantly telling us that some things don't work perfectly today, and asking for quick shortcuts isn't helping.

> They dont work for testing apps, running prerelease software and a lot more.

This is just not true.  GNOME [offers](https://nightly.gnome.org/) a Flatpak repository for testing nightly builds of GNOME applications.  If you are outside the GNOME umbrella, then you can use [Flathub Beta](https://docs.flathub.org/docs/for-users/installation/) to publish pre-release snapshots.

> There are also occasions where there is no flatpak. For sure this is not a good scenario, but in the case of immutable Distros, many people need containers.

Do you have specific examples where creating a Flatpak for a graphical application would be more difficult than creating a distribution package?  I am curious.

> To the comment above, I dont know where you read that installing packages at creation would be needed.

That's what Distrobox does, because that's what you need to do if you advertise support for random generic OCI images.
